src/transformers/models/bert/modeling_tf_bert.py

```diff
@@ -347,15 +241,21 @@ def call(
         output_attentions: bool,
         training: bool = False,
     ) -> Tuple[tf.Tensor]:
-        query_layer = self.query(inputs=hidden_states)
-        key_layer = self.key(inputs=hidden_states)
-        value_layer = self.value(inputs=hidden_states)
+        batch_size = shape_list(hidden_states)[0]
+        mixed_query_layer = self.query(inputs=hidden_states)
+        mixed_key_layer = self.key(inputs=hidden_states)
+        mixed_value_layer = self.value(inputs=hidden_states)
+        query_layer = self.transpose_for_scores(mixed_query_layer, batch_size)
+        key_layer = self.transpose_for_scores(mixed_key_layer, batch_size)
+        value_layer = self.transpose_for_scores(mixed_value_layer, batch_size)
 
         # Take the dot product between "query" and "key" to get the raw
         # attention scores.
+        # (batch size, num_heads, seq_len_q, seq_len_k)
+        attention_scores = tf.matmul(query_layer, key_layer, transpose_b=True)
         dk = tf.cast(self.attention_head_size, dtype=query_layer.dtype)
-        query_layer = tf.multiply(query_layer, tf.math.rsqrt(dk))
-        attention_scores = tf.einsum("aecd,abcd->acbe", key_layer, query_layer)
+        one_cst = tf.constant(1.0, dtype=dk.dtype)
+        query_layer = tf.multiply(query_layer, tf.divide(one_cst, tf.math.sqrt(dk)))
```

Comment:
Cache this `1.0 / sqrt(...)` in the initialization method or build? 

This doesn't change in between calls, so you save some time not re-computing.

Reply:
This indeed can be put in the init, I will update this 👍  But cannot be done in plain Python, because the dtype has to be compliant in case of mixed precision defined to something else than `float32`.

Reply:
Can't we compute over plain old python and store the results _(in `init`)_ with `tf.constant(..., dtype=...)`

Reply:
We can absolutely doing that 👍 We can detect any policy update with `self._compute_dtype` in a layer.

Reply:
Ah finally no, I found a use case for which it fails. If you run `tf.keras.mixed_precision.experimental.set_policy('mixed_float16')` before to instantiate your model, it works, if you do it after it fails. So it is safer to have the cast inside the `call`.